### ISSUE TITLE
Language guidelines

### ISSUE THREAD

**Issue by efrecon** [eclipse/mosquitto#800]
I have just submitted PR #799 for the generation of Docker images for all official releases of mosquitto. I also wanted to address issue #510, i.e. provide a way for users of this image to configure it from the outside using environment variables instead of using one or several configuration files. I have already done all necessary work in a separate [project](https://github.com/efrecon/docker-mosquitto). However, this requires to split the (default) configuration file into several subsections in order for the environment variables to be able to address each configuration option in a unique way. Splitting the code into sections is implemented using Tcl, my goto-language for automation and scripting. I have seen that testing of the project relies on Python. Do you have any guidelines as to which languages you are willing to accept as part of the project? This would only concern the internals of the Docker image, so it is well-contained.

**Comment by toast-uz** [comment]
As my understanding, mosquitto has no code depended on a specific language except for LANG=C. Does this become an answer of your question?

**Comment by karlp** [comment]
I think it's more a question for the docker community.  I personally would have thought that tcl would be considered akin to perl (aka, what sort of obsolete nonsense is this?!) to the docker community.  But hey, I also don't see why you need to do a completely different way of configuring things, surely you can just put your own config into your own container? isn't that what they're for?  But, I'm not a docker user, so *shrugs*

Generally though, python2, and these days, python3 are available in the default installations in most places.  Tcl isn't.  That should provide another useful data point.

**Comment by efrecon** [comment]
There are two questions to address here I think:
1. Should  mosquitto provide for easier configuration of its instance within a Docker container using environment variables?
2. If so, how should this be implemented? (which is where the language question gets its place).

As for question 1 above, I think that providing easy configuration through environment variables is inline with what most other Docker images provide. Take the official [postgres](https://hub.docker.com/_/postgres/) for example, you'll see that it goes a long way providing quick setup using environment variables, while still providing ways for experienced people to tune the configuration using files. This is also true for [mysql](https://hub.docker.com/_/mysql/), the official [registry](https://docs.docker.com/registry/deploying/), etc.

Question 2 is probably open for debate. However, the Dockerfile builds on top of Alpine linux, which is a minimal distribution. In the raw Alpine Docker image, there isn't any bash, perl, python2/3, tcl or any other dynamic language, you'll have to add the one(s) you need. So I guess that the point gets back to is the requirement associated to question 1 strong enough to justify an implementation, and if yes, how should this implementation look like? I picked Tcl because it is what I am most familiar with and because it's been very stable for a number of years. Reimplementing the splitting of the configuration file into bits and pieces using [ash](https://linux.die.net/man/1/ash), the default shell in Alpine is certainly possible, but would require an even larger effort.

I don't want to do a PR if I know it's going to be rejected, there isn't any point. This was the main point of this thread. It all boils down to whether the project think that configuration through environment variables is a good idea or not. I think it is, and is inline with what the rest of the Docker community does.

**Comment by karlp** [comment]
Personally I don think there's any sane way of configuring things via environment variables in any manner that's remotely manageable.  Surely you have orchestration of your containers anyway right?  You've got lists of topics to bridge, lists of bridges, none of this goes well with environment variables.  I _realllly_ don't see what doing any of this config via a sea of environment variables brings to the table, but then, again, I don't use docker.

**Comment by Assar63** [comment]
If environment variable are desired there are solutions. One way of using environment variables is that when starting the MQTT docker container, just before starting mosquitto, populate the mosquitto configuration file. This way it's possible to use the same configuration template among multiple environments. For instance when using a docker stack for DEV, INTEGRATION, TEST, STAGE and PROD.
Mosquitto should not not need to deal with specifics on special target environments.

**Comment by efrecon** [comment]
@Assar63 You are entirely right and this is exactly what the [separate project](https://github.com/efrecon/docker-mosquitto) that I propose to include as a PR does. Note that populating the configuration via environment variables is not a straightforward task as configuration variables in the file appear several times under the same name, but with a different purpose. For example, if you wanted to setup a regular MQTT listener and a websocket listener, there is no straightforward way to address the configuration value called `protocol` in the configuration file, as it appears (at least) twice, once for the main listener and a second time for the additional listener, meaning you cannot map the content of a variable called `MOSQUITTO_PROTOCOL` (as is the de-facto standard way of doing in Docker images) to set the variable `protocol`: you wouldn't know which one to set in a deterministic way. This very specific problem is what my Tcl script and proposed solution is trying to address.

Reading the [entrypoint](https://github.com/efrecon/docker-mosquitto/blob/master/docker-entrypoint.sh) implementation should provide you with a better understanding of the problem as it performs environment variable substitution twice, the second time based on those virtual "sections" of the configuration file.

**Comment by efrecon** [comment]
@Assar63 , you are perfectly right. The de-facto standard in Docker seems to be environment variables for simple configuration, shared files between host/swarm and container for more complex cases. Currently, mosquitto only offers configuration through files when used in Docker environments.

I am not proposing to remove the ability to configure through files. Just adding ways to properly configure using environment variables.

**Comment by efrecon** [comment]
I suppose that the main use case is to be able to tighten up several containers as quickly as possible using [compose](https://docs.docker.com/compose/compose-file/) or stack files, without having to delve into the internals of each technology and especially how it can be configured. Somebody else (image maintainers) have done the work of extracting and sorting out which configuration parameters are most likely to be modified and exposed these configuration as environment variables. Again, this the way the most prominent Docker images are built up: [postgres](https://hub.docker.com/_/postgres/), [mysql](https://hub.docker.com/_/mysql/), [RabbitMQ](https://hub.docker.com/_/rabbitmq/), [InfluxDB](https://hub.docker.com/_/influxdb/), [grafana](https://hub.docker.com/r/grafana/grafana/) and a whole lot more have this duality when it comes to configuration. I am pretty sure there are arguments against having two different ways to configure the same thing, but it seems to anyhow be what Docker users expect from images on the store/hub.

**Comment by toast-uz** [comment]
Closing because this place is not for a question. Look at the instruction (https://mosquitto.org/).
